Minio/DataModel/Select/SelectResponseStream.cs

```diff
@@ -0,0 +1,249 @@
+using System.Xml.Serialization;
+using RestSharp;
+using System;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Force.Crc32;
+using System.Linq;
+using System.Collections.Generic;
+
+namespace Minio.DataModel
+{
+    [Serializable]
+    public class SelectResponseStream
+    {
+        // SelectResponseStream is a struct for selectobjectcontent response.
+        
+        public Stream Payload  { get ; set; }  
+
+        [XmlElement("Stats", IsNullable = false)]
+        public StatsMessage Stats { get; set; }
+
+        [XmlElement("Progress", IsNullable = false)]
+        public ProgressMessage Progress { get; set; }
+
+        private byte[] prelude = new byte[8];
+        private byte[] preludeCRC = new byte[4];
+        private byte[] messageCRC = new byte[4];
+        private byte[] headerValueLen = new byte[2];
+
+        private MemoryStream payloadStream = null;
+        private bool _isProcessing;
+        
+        public SelectResponseStream()
+        {
+        }
+        public static void CopyToStream(Stream input, Stream output)
+        {
+            byte[] buffer = new byte[16*1024];
+            int read;
+            while((read = input.Read (buffer, 0, buffer.Length)) > 0)
+            {
+                output.Write (buffer, 0, read);
+            }
+            output.Seek(0, SeekOrigin.Begin);
+        }
+
+        protected int ReadFromStream(byte[] buffer)
+        {
+            int read = -1;
+            if  (!this._isProcessing)
+            {
+               return read;
+            }
+            read = this.payloadStream.Read(buffer, 0, buffer.Length);
+            if (!this.payloadStream.CanRead)
+            {
+                this._isProcessing = false;
+            }
+            return read;
+        }
+        public SelectResponseStream(Stream s)
+        {
+            if (s != null) {
+                var _ms = new MemoryStream();
+                CopyToStream(s, _ms);
+                this.payloadStream = _ms;
+                this.Payload = new MemoryStream();
+            }
+            this._isProcessing = true;
+            this.payloadStream.Seek(0, SeekOrigin.Begin);
+            this.start();
+        }
+        
+        private void start()
+        {
+            int numBytesRead = 0;
+            int msgOffset = 0;
+            while (_isProcessing)
+            {   
+                var n =  ReadFromStream(prelude);
+                numBytesRead += n;
+                n =  ReadFromStream(preludeCRC);
+                numBytesRead += n;
+                var inputArray = new byte[prelude.Length + preludeCRC.Length];
+                System.Buffer.BlockCopy(prelude,0,inputArray,0, prelude.Length);
+                System.Buffer.BlockCopy(preludeCRC,0,inputArray,prelude.Length, preludeCRC.Length);
+
+                // write real data to inputArray
+                Crc32Algorithm.ComputeAndWriteToEnd(inputArray); // last 4 bytes contains CRC
```

Comment:
@poornas We are actually overwriting the last 4 bytes (preludeCRC) with the calculated CRC in this function `ComputeAndWriteToEnd`

```
        /// <summary>
        /// Computes CRC-32 from input buffer and writes it after end of data (buffer should have 4 bytes reserved space for it). Can be used in conjunction with <see cref="IsValidWithCrcAtEnd(byte[],int,int)"/>
        /// </summary>
        /// <param name="input">Input buffer with data to be checksummed.</param>
        /// <param name="offset">Offset of the input data within the buffer.</param>
        /// <param name="length">Length of the input data in the buffer.</param>
        /// <returns>CRC-32 of the data in the buffer.</returns>
        public static uint ComputeAndWriteToEnd(byte[] input, int offset, int length)
        {
            if (length + 4 > input.Length)
                throw new ArgumentOutOfRangeException("length", "Length of data should be less than array length - 4 bytes of CRC data");
            var crc = Append(0, input, offset, length);
            var r = offset + length;
            input[r] = (byte)crc;
            input[r + 1] = (byte)(crc >> 8);
            input[r + 2] = (byte)(crc >> 16);
            input[r + 3] = (byte)(crc >> 24);
            return crc;
        }
 
        /// <summary>
        /// Computes CRC-32 from input buffer - 4 bytes and writes it as last 4 bytes of buffer. Can be used in conjunction with <see cref="IsValidWithCrcAtEnd(byte[])"/>
        /// </summary>
        /// <param name="input">Input buffer with data to be checksummed.</param>
        /// <returns>CRC-32 of the data in the buffer.</returns>
        public static uint ComputeAndWriteToEnd(byte[] input)
        {
            if (input.Length < 4)
                throw new ArgumentOutOfRangeException("input", "input array should be 4 bytes at least");
            return ComputeAndWriteToEnd(input, 0, input.Length - 4);
        }
```
Our `inputArray` here which is of 12bytes length has 8bytes(prelude)+4(preludeCRC) and this function `ComputeAndWriteToEnd` will calculate the CRC for (n - tailing 4) bytes and write the CRC to the last 4 bytes in inputArray. Which will overwrite the preludeCRC.

I think we should not copy the preludeCRC before validating. ie, We can do something like

```
var inputArray = new byte[prelude.Length + 4]; // 4 is the space for storing the calculated CRC
System.Buffer.BlockCopy(prelude,0,inputArray,0, prelude.Length); // Copy the prelude only
Crc32Algorithm.ComputeAndWriteToEnd(inputArray); 
// This will calculate the CRC for the prelude (first 8 bytes of the inputArray) and write the
// calculated CRC in the last 4 bytes of the inputArray
..
..
// We can then check if the last 4 bytes match the `preludeCRC` obtained from the input stream.
// (OR)
// We could calculate the crc by `var crc = Append(0, input, offset, length);` and then check if it 
// matches preludeCRC from the input stream.
```

Reply:
good catch - fixed this